### PR TITLE
Editing ability

### PR DESCRIPTION
**Pull request by YewoMhango**
### Some of the major additions:

- Ability to edit the imported songs using a WYSIWYG editor (using Tiptap) or editing the raw code
- Dark/light mode

### I'm also considering adding these features later:

- Ability to start a new hymnal if you don't already have a JSON file
- Saving edit progress to localStorage
- Diffing like GitHub to compare changes with the original

A live version can also be viewed here: https://cis-previewer.vercel.app/

### CODE REVIEW

File: src/components/HymnEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+}) {
+    if (!currentHymn) {
+        return (
+            <Center>
+                <NoHymnMessage />
+            </Center>
+        );
+    }
+
+    let [editingRawCode, setEditingRawCode] = useState(false);
```
Comment (OlivierJM) [inline src/components/HymnEditor.tsx]: If this doesn't need to be a "let" let's change it `const`

File: src/components/HymnEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+            </Center>
+        );
+    }
+
+    const getEditorContent = (editor: Editor) =>
+        hymn?.format === "html"
+            ? editor?.getHTML()
+            : (editor?.storage.markdown.getMarkdown() as string);
+
+    let delay = useDebouncedAction();
```
Comment (OlivierJM) [inline src/components/HymnEditor.tsx]: same here.

File: src/components/HymnEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+                        editor?.commands.setContent(text);
+                        updateHymnContent(text);
+                    }}
+                />
+            )}
+        </Container>
+    );
+}
+
+function TitleInput({
```
Comment (OlivierJM) [inline src/components/HymnEditor.tsx]: Should we maybe have this in its own files ?

File: src/components/HymnEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    hymnNumber,
+    title,
+    updateTitle,
+}: {
+    title: string;
+    hymnNumber: number;
+    updateTitle: (title: string) => void;
+}) {
+    let [value, setValue] = useState(title);
```
Comment (OlivierJM) [inline src/components/HymnEditor.tsx]: this too should be a `const`

File: src/components/HymnEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+            <RichTextEditor.Content
+                style={{
+                    fontFamily: "'Nunito', var(--mantine-font-family)",
+                }}
+            />
+        </RichTextEditor>
+    );
+}
+
+function RawCodeEditor({
```
Comment (OlivierJM) [inline src/components/HymnEditor.tsx]: Maybe this as well we can have it in its own file.

**Review by OlivierJM — COMMENTED** [review]
Overall this looks great, The initial one was done in a hurry but this adds lots of improvements. 
Let's try and separate files for better management in case we add more features.

**Comment by OlivierJM** [comment]
@YewoMhango We can also include in the ReadMe that there's a editing capability.

File: src/Context/ColorScheme.tsx [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+type ColorSchemeContextType = {
+    colorScheme: "dark" | "light";
+    toggleColorScheme: () => void;
+};
+
+const ColorSchemeContext = createContext<ColorSchemeContextType | undefined>(
+    undefined
+);
+
+export function ColorSchemeProvider({ children }: { children: ReactNode }) {
```
Comment (OlivierJM) [inline src/Context/ColorScheme.tsx]: There is another project where I use something like this for managing the color scheme in a context. 

```js
 const prefersDarkMode = useMediaQuery('(prefers-color-scheme: dark)'); // default to user's current color scheme ==> window.matchMedia("(prefers-color-scheme: dark)").matches

  const [mode, setMode] = useState<PaletteMode>(prefersDarkMode ? 'dark' : 'light');

  const handleTheme = () => {
    setMode((prevMode) => (prevMode === 'light' ? 'dark' : 'light'));
  };

```

This you shouldn't need a useEffect.

**Comment by YewoMhango** [comment]
Thanks a lot for the comments

File: src/Context/ColorScheme.tsx (hunk 6 above)
Comment (YewoMhango) [reply src/Context/ColorScheme.tsx]: Thanks a lot. The Mantine hook really simplifies it a lot

But also, the `useEffect` part is actually there to update the color scheme if the user changes their system color preference after the app is already opened. With your other suggestion, I've currently simplified the code to this:

```ts
const prefersDarkMode = useMediaQuery("(prefers-color-scheme: dark)");
const [colorMode, setColorMode] = useState<"dark" | "light">(
    prefersDarkMode ? "dark" : "light"
);

useEffect(() => {
    setColorMode(prefersDarkMode ? "dark" : "light");
}, [prefersDarkMode]);

const toggleColorMode = () => {
    setColorMode((prevScheme) =>
        prevScheme === "light" ? "dark" : "light"
    );
};
```

But regardless, we could still remove it if you think it's unnecessary or disruptive

**Review by OlivierJM — APPROVED** [review]

**Comment by OlivierJM** [comment]
This is great work  🚀 🙌🏽

**Pull request merged**